src/utils/display-observer.js

```diff
@@ -0,0 +1,248 @@
+  /**
+   * @param {!AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.docObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, DOC_OBSERVER),
+      {
+        root: ampdoc.getBody(),
+        threshold: 0.001,
+      }
+    );
+
+    /** @private @const {!IntersectionObserver} */
+    this.viewportObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, VIEWPORT_OBSERVER),
+      {threshold: 0.001}
+    );
+
+    /** @private {boolean} */
+    this.isDocVisible_ = ampdoc.isVisible();
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() => {
+      if (ampdoc.isVisible() !== this.isDocVisible_) {
+        this.isDocVisible_ = ampdoc.isVisible();
+        this.docVisibilityChanged_();
+      }
+    });
+
+    /** @private @const {!Map<!Element, !Array<function(boolean)>>} */
+    this.targetObserverMultimap_ = new Map();
+
+    /** @private @const {!Map<!Element, !Array<boolean>>} */
+    this.targetDisplayMap_ = new Map();
+  }
+
+  /** @override */
+  dispose() {
+    this.docObserver_.disconnect();
+    this.viewportObserver_.disconnect();
+    this.visibilityUnlisten_();
+    this.visibilityUnlisten_ = null;
+    this.targetObserverMultimap_.clear();
+    this.targetDisplayMap_.clear();
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function(boolean)} callback
+   */
+  observe(target, callback) {
+    let callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      callbacks = [];
+      this.targetObserverMultimap_.set(target, callbacks);
+      this.docObserver_.observe(target);
+      this.viewportObserver_.observe(target);
+    }
+    if (pushIfNotExist(callbacks, callback)) {
+      if (this.targetDisplayMap_.has(target)) {
+        setTimeout(() => {
+          const display = getDisplay(
+            this.targetDisplayMap_.get(target),
+            this.isDocVisible_
+          );
+          if (display != null) {
+            callCallbackNoInline(callback, display);
+          }
+        });
+      }
+    }
+  }
+
+  /**
+   * @param {!Element} target
+   * @param {function()} callback
+   */
+  unobserve(target, callback) {
+    const callbacks = this.targetObserverMultimap_.get(target);
+    if (!callbacks) {
+      return;
+    }
+    removeItem(callbacks, callback);
+    if (callbacks.length == 0) {
+      this.targetObserverMultimap_.delete(target);
+      this.targetDisplayMap_.delete(target);
+      this.docObserver_.unobserve(target);
+      this.viewportObserver_.unobserve(target);
+    }
+  }
+
+  /** @private */
+  docVisibilityChanged_() {
+    this.targetObserverMultimap_.forEach((callbacks, target) => {
+      const displays = this.targetDisplayMap_.get(target);
+      const oldDisplay = getDisplay(displays, !this.isDocVisible_);
+      const newDisplay = getDisplay(displays, this.isDocVisible_);
```

Comment:
This double call to `getDisplay` seems really weird. I think we're trying to prevent a double `isDisplayed = true` (or `false`) call from happening? If so, it seems like we should be storying the last display result for each element in the mapping.

Reply:
I did a bunch of renaming and I think that should make it much more readable. In a nutshell:

1. The observers produce observations (renamed all structures to `observations`)
2. The `computeDisplay()` takes observations and document's visibility to compute the actual `isDisplayed` signal.
3. Based on this, the display computation would produce different results for different observations and document visibility values. The callbacks are called only when the actual display value has changed.

PTAL if this new notation clarifies the code sufficiently.